src/datasets/arrow_writer.py

```diff
@@ -293,18 +311,48 @@ def write_rows_on_file(self):
         self.write_table(table)
         self.current_rows = []
 
-    def write(self, example: Dict[str, Any], writer_batch_size: Optional[int] = None):
-        """Add a given Example to the write-pool of examples which is written to file.
+    def write(
+        self,
+        example: Dict[str, Any],
+        key: Optional[Union[str, int, bytes]] = None,
+        writer_batch_size: Optional[int] = None,
+    ):
+        """Add a given (Example,Key) pair to the write-pool of examples which is written to file.
 
         Args:
             example: the Example to add.
+            key: Optional, a unique identifier(str, int or bytes) associated with each example
         """
-        self.current_examples.append(example)
+        # Utilize the keys and duplicate checking when `self._check_duplicates` is passed True
+        if self._check_duplicates:
+            # Create unique hash from key and store as (key, example) pairs
+            hash = self._hasher.hash(key)
+            self.current_examples.append((example, hash))
+            # Maintain record of keys and their respective hashes for checking duplicates
+            self.hkey_record.append((hash, key))
+        else:
+            # Store example as a tuple so as to keep the structure of `self.current_examples` uniform
+            self.current_examples.append((example,))
```

Comment:
To follow the typing `Tuple[Dict[str, Any], str]`
```suggestion
            self.current_examples.append((example, ""))
```